posts/java-enums-with-extra-visitor-please.md

```diff
@@ -0,0 +1,203 @@
+---
+authors:
+- Grégory ELHAIMER
+tags:
+- Java
+- Software Development
+- Software Craftmanship
+date: 2019-05-28T12:21:50.000Z
+title: "Java enums? With extra Visitor please!"
+image: 
+---
+
+Enums can be seen as a group of strongly typed constants. Their several use cases include: attributing strong semantics to values, limiting and validating available values for data, and enhancing code readability.
+Using enums can be tricky when it comes to making decisions according to its values. As the enum evolves, each part of the code where it has been used as a condition must be checked. If many decisions have been based on an enum's values, maintaining the code base can be a nightmare. A single missed verification can lead to corruption of the entire system.
+
+Is there a way to reduce the impact of enums' impact on the code base?
+
+# The switch-case approach
+
+Consider the `AssetClass` enum which represents distinct groups of tradable commodities:
+
+```java
+public enum AssetClass {
+    METAL,
+    ENERGY,
+    AGRICULTURAL,
+}
+```
+
+This enum can be used to decide which strategy, mapper or any other behavior to use according to the value of `AssetClass`. A general example of this is given in the following code:
+
+```java
+public AutomatedTradingStrategy getAutomatedTradingStrategy(AssetClass assetClass) {
+    switch (assetClass) {
+        case METAL:return new HedgingStrategy();
+        case ENERGY: return new SwingTradingStrategy();
+        case AGRICULTURAL:
+        default: return DayTradingStrategy();
+    }
+}
+```
+
+The switch-case statement is probably the most straightforward way to do this. However, it has several flaws.
+
+The method `getAutomatedTradingStrategy()` returns a behavior according to the value of `AssetClass`. Defining a default behavior then becomes mandatory, even if in this example all the values of `AssetClass` enum are handled. We can do this by either returning a default implementation of `AutomatedTradingStrategy`, null, or throwing an exception.
+
+Even though each enum value should have a handler adding a default handler here ensures that there is some way to handle an arbitrary `AssetClass`.  This is helpful when an enum value has been added but is missed in other parts of the codebase.
```

Comment:
Even though each enum value should have a handler, adding a default handler here ensures that there is some way to handle an arbitrary `AssetClass`.